tests/components/broadlink/test_discovery.py

```diff
@@ -0,0 +1,175 @@
+
+
+async def test_discovery_update_ip_address(hass):
+    """Test we update the entry when a known device is discovered with a different IP address."""
+    device = get_device("Living Room")
+    _, mock_entry, _ = await device.setup_entry(hass)
+
+    previous_host = device.host
+    device.host = "192.168.1.128"
+    results = device.get_mock_discovery() * 2
+
+    with device.patch_setup(mock_discovery=results), patch(
+        "homeassistant.components.broadlink.helpers.socket.gethostbyname",
+        return_value=previous_host,
+    ) as mock_host:
+        await hass.data[DOMAIN].discovery.coordinator.async_refresh()
+        await hass.async_block_till_done()
+
+    assert mock_host.call_count == 2
+    assert mock_entry.data["host"] == device.host
+
+
+async def test_discovery_update_hostname(hass):
+    """Test we update the entry when the hostname is no longer valid."""
+    device = get_device("Living Room")
+    results = device.get_mock_discovery()
+    device.host = "invalidhostname"
+
+    _, mock_entry, _ = await device.setup_entry(hass, mock_discovery=results)
+
+    device.host = "192.168.1.128"
+    results = device.get_mock_discovery() * 2
+
+    with device.patch_setup(mock_discovery=results), patch(
+        "homeassistant.components.broadlink.helpers.socket.gethostbyname",
+        side_effect=OSError(socket.EAI_NONAME, None),
+    ) as mock_host:
+        await hass.data[DOMAIN].discovery.coordinator.async_refresh()
+        await hass.async_block_till_done()
+
+    assert mock_host.call_count == 2
+    assert mock_entry.data["host"] == device.host
+
+
+async def test_discovery_do_not_change_hostname(hass):
+    """Test we do not update the entry if the hostname routes to the device."""
+    device = get_device("Living Room")
+    results = device.get_mock_discovery()
+    device.host = "somethingthatworks"
+
+    _, mock_entry, _ = await device.setup_entry(hass, mock_discovery=results)
+
+    with device.patch_setup(mock_discovery=results), patch(
+        "homeassistant.components.broadlink.helpers.socket.gethostbyname",
+        return_value=device.host,
+    ) as mock_host:
+        await hass.data[DOMAIN].discovery.coordinator.async_refresh()
+        await hass.async_block_till_done()
+
+    assert mock_host.call_count == 1
+    assert mock_entry.data["host"] == "somethingthatworks"
+
+
+async def test_discovery_ignore_os_error(hass):
+    """Test we do not propagate an OS error during a discovery."""
+    device = get_device("Living Room")
+    _, mock_entry, _ = await device.setup_entry(hass)
+
+    with patch.object(hass.config_entries.flow, "async_init") as mock_init, patch(
+        "homeassistant.components.broadlink.get_broadcast_addrs",
+        return_value=["192.168.31.0"],
+    ), patch(
+        "homeassistant.components.broadlink.discovery.blk.xdiscover",
+        side_effect=OSError(),
+    ) as mock_discovery:
+        try:
+            await hass.data[DOMAIN].discovery.coordinator.async_refresh()
```

Comment:
We want to avoid interacting with integration details in the tests. Can we set up the integration and provoke the same behavior with core interfaces or by moving time forward?

Eg we have:
https://github.com/home-assistant/core/blob/2741362b23a35e3806384882b61374af1e591957/tests/common.py#L287-L289